src/mgr/Mgr.cc

```diff
@@ -95,56 +95,58 @@ class MetadataUpdate : public Context
   void finish(int r) override
   {
     daemon_state.clear_updating(key);
-    if (r == 0) {
-      if (key.first == "mds") {
-        json_spirit::mValue json_result;
-        bool read_ok = json_spirit::read(
-            outbl.to_str(), json_result);
-        if (!read_ok) {
-          dout(1) << "mon returned invalid JSON for "
-                  << key.first << "." << key.second << dendl;
-          return;
-        }
 
-        json_spirit::mObject daemon_meta = json_result.get_obj();
+    if (r) {
+      dout(1) << "mon failed to return metadata for "
+              << key.first << "." << key.second << ": "
+	      << cpp_strerror(r) << dendl;
+      return; 
+    }
 
-        // Apply any defaults
-        for (const auto &i : defaults) {
-          if (daemon_meta.find(i.first) == daemon_meta.end()) {
-            daemon_meta[i.first] = i.second;
-          }
-        }
+    if (key.first != "mds" && key.first != "osd") {
+      dout(1) << "unknown service"
+              << key.first << "." << key.second <<  dendl;
+      ceph_abort();
+    }
 
-        DaemonStatePtr state;
-        if (daemon_state.exists(key)) {
-          state = daemon_state.get(key);
-	  Mutex::Locker l(state->lock);
-          daemon_meta.erase("name");
-          daemon_meta.erase("hostname");
-          state->metadata.clear();
-          for (const auto &i : daemon_meta) {
-            state->metadata[i.first] = i.second.get_str();
-          }
-        } else {
-          state = std::make_shared<DaemonState>(daemon_state.types);
-          state->key = key;
-          state->hostname = daemon_meta.at("hostname").get_str();
+    json_spirit::mValue json_result;
+    bool read_ok = json_spirit::read(outbl.to_str(), json_result);
+    if (!read_ok) {
+      dout(1) << "mon returned invalid JSON for "
+              << key.first << "." << key.second << dendl;
+      return;
+    }
 
-          for (const auto &i : daemon_meta) {
-            state->metadata[i.first] = i.second.get_str();
-          }
+    json_spirit::mObject daemon_meta = json_result.get_obj();
 
-          daemon_state.insert(state);
-        }
-      } else if (key.first == "osd") {
-      } else {
-        ceph_abort();
+    if (daemon_meta.count("hostname") == 0) {
+      dout(1) << "Skipping incomplete metadata entry" << dendl;
+      return;
+    }
+
+    // Apply any defaults
+    for (const auto &i : defaults) {
+      if (daemon_meta.find(i.first) == daemon_meta.end()) {
+        daemon_meta[i.first] = i.second;
       }
+    }
+
+
+    DaemonStatePtr state;
```

Comment:
Because DaemonState contains things like a short history of perf counters, it is not desirable to create a new instance whenever the metadata updates.  We should be checking for an existing instance and update the metadata on that, like the existing code (sort of) did

Reply:
We clear the metadata (map of strings) before re-populating it from the update, but not the whole DaemonState (everything else, including perf counters) object

Reply:
thank you very much.